src/libopenrave-core/environment-core.h

```diff
@@ -1639,6 +1640,22 @@ class Environment : public EnvironmentBase
         return _mutexEnvironment;
     }
 
+    virtual void IterateBodies(const std::function<void(const KinBodyPtr&)>& mapFunction) override
+    {
+        // Add one to the active iterator count, and release our read when exiting.
+        // This reader count allows us to catch abuses of the iterator API (mutation during iteration) that might otherwise cause undefined behaviour.
+        _vecbodiesActiveReaders.fetch_add(1, std::memory_order::memory_order_acq_rel);
+        std::shared_ptr<void> __deferReleaseRead{
+            nullptr, [&](nullptr_t) { _vecbodiesActiveReaders.fetch_sub(1, std::memory_order::memory_order_acq_rel); }};
+
+        // Map the provided function over all of the live bodies in the environment
+        for (const KinBodyPtr& pBody : _vecbodies) {
```

Comment:
is this safe without locking `_mutexInterfaces`? @kanbouchou can you take a look?